dawn/docs/parallel_model.rst

```diff
@@ -0,0 +1,251 @@
+
+Formal Definition of the Dawn Parallel Model
+============================================
+
+Contract to the user from HIR
+-----------------------------
+
+Dawn will generate code based on the operations passed from high-level
+intermediate representation. It's specification can be found `here`_.
+
+Dependencies
+~~~~~~~~~~~~
+
+Generally, stencil computation in dawn differentiates between horizontal
+and the vertical dimension. The assumption is that individual statements
+of a stencil are embarrassingly parallel in the horizontal. For that
+reason, there is no guarantee that any horizontal point's computation
+will be finished before any other in each user statement.
+
+Vertical Dependencies
+^^^^^^^^^^^^^^^^^^^^^
+
+Each vertical region of a stencil specifies a loop order of execution.
+Dawn will uphold the following rules:
+
+-  If the loop order is forward, any statement with a vertical
+   dependency on other statements executed on a vertical level with
+   index ``k`` will be executed after each statement that it depends on
+   that appears previously in the current vertical region or in any
+   previous vertical region was executed on every vertical level with
+   indices ``k-1`` and smaller.
+-  If the loop order is backward, any statement with a vertical
+   dependency on another statement executed on a vertical level with
+   index ``k`` will be executed after each statement that it depends on
+   that appears previously in the current vertical region or in any
+   previous vertical region was executed on every vertical level with
+   indices ``k+1`` and bigger.
```

Comment:
Can you rephrase this sentence? I had trouble parsing this sentence... 
1. "after each statement that it depends on": If we say that we can implement the toolchain in any way if the code runs _as if it would obey this rules_, we could just skip this part and say "after each statement that appears".
2. "all previous statements were executed on every vertical level with indices ``k-1`` and smaller". Now think about
```
    vertical_region(k_start, k_end)
      one = zero[k]
    vertical_region(k_start, k_end)
      two = one[k + 1]
```
If I understand correctly from what you write, this would behave the same way as
```
    vertical_region(k_start, k_end)
      one = zero[k]
      two = one[k + 1]
```
But it should not, because it should execute the whole first vertical_region first (but because of the "as-if" rule, the toolchain might do something like
```
    vertical_region(k_start, k_end)
      one = zero[k + 1]
      two = one[k]
``

Reply:
1) sounds good

2) we need to define a target here, our current implementation creates the same result (one backward MS with both lines in the same do).

Reply:
I didnt understand Lukas point 2, but description with the point 1 Lukas sounds good.

Reply:
Let me explain again point 2:
The user can have
```
    vertical_region(k_start, k_end)
      one = zero
    vertical_region(k_start, k_end)
      two = one[k + 1]
```
What will be in variable two in the end? It might be `zero[k + 1]` or it might be the original `one[k + 1]`.

I personally might be tempted to think that first, `one = zero[k]` is applied to the whole vertical region, and after that, the `two = one[k + 1]` is applied to the whole vertical region, because we have two vertical regions (and they have a direction, a start, and an end).

But as far as I understand, the text says (with some parentheses for my own understanding ...):

> any [statement with a vertical dependency on another statement executed on a vertical level with index ``k``] will be executed [after each statement that it depends on that appears ({previously in the current vertical region} or {in any previous vertical region}) was executed on every vertical level with indices ``k-1`` and smaller].

So it says that all statements in a previous vertical region are executed on every vertical level with indices `k-1` and smaller. Which is equivalent to write:

```
    vertical_region(k_start, k_end)
      one = zero
      two = one[k + 1]
```
So for each level, both statements are applied, and then we move to the next level.

Reply:
Is that the intended behaviour?
If the loop order is forward, any statement executed on a vertical level with index ``k`` will be executed after each statement that appears in the current vertical region on every vertical level with indices ``k-1`` and smaller.
Each previous vertical region was executed on its full iteration space.

Reply:
What should happen here:
```
vertical_region(k_start, k_end):
    one = 0
vertical_region(k_start, k_end):
    one = 1
    two = one[k + 1] (1) + one (2) + one[k - 1] (3)
    one = 2
    two = two + one[k + 1] (4) + one (5) + one[k - 1] (6)
```
What is happening in dawn:
```
Stencil_0
    MultiStage_0
        Stage_0
            Do
                one = 0
        Stage_1
            Do
                one = 1
    MultiStage_1
        Stage_0
            Do
                two = one[k + 1] (1) + one (2) + one[k - 1] (3)
    MultiStage_2
        Stage_0
            Do
                one = 2
    MultiStage_3
        Stage_0
            Do
                two = two + one[k + 1] (4) + one (5) + one[k - 1] (6)
```
So if I understand correctly, it means that (1), (2) and (3) refer to `one = 1`, while (4), (5) and (6) refer to `one = 2`.

Is this really what we expect our parallel model to do?

Reply:
@cosunae ?

Reply:
I need some help in understanding the statements (sorry, but I'm slow). Taking Lukas's example:

```
vertical_region(k_start, k_end)
     one = zero
vertical_region(k_start, k_end)
      two = one[k + 1]
```
If the two vertical_regions are just parallel ranges, and if I read "the first vertical statement is executed before the second one" the loop could be transformed into
```
vertical_region(k_end, k_start)
      one = zero[k+1]
```
What's wrong with my reasoning?

Reply:
so the goal was to not express parallelism in the front end (and hence the SIR). On an easy level this means 
```
vertical_region(k_start, k_end)
  a = b
vertical_region(k_start, k_end)
  c = d
```
should yield to the same output code as 
```
vertical_region(k_start, k_end)
  a = b
  c = d
```
This can be achieved. 
The question is really what we expect to happen with k-dependencies. Here we can change the behaviour of the parallelizer if we have a clear idea what we promise the user. 
There are two different computations you can express:

```
  vertical_region(k_start, k_end)
    one = zero
  vertical_region(k_start, k_end)
    two = one[k + 1]
````
can mean set all of one to zero and then set all of two to the updated one. The result of this would be `two=zero` and `one = zero`. Where 
```
  vertical_region(k_start, k_end)
    one = zero
    two = one[k + 1]
````
can be interpreted as update my current one and write the old one to two. The result of this would be `one = zero` `two=one`. 

The current implementation of dawn does not distinguish between the two and always goes for the second option. We need to clarify if these should be different things though.

And if they are different (my opinion is they should be since there are two motifs to express) we need to think very carefully about how Lukas' last example should be interpreted. 

I hope this clarifies things for you @mbianco?

Reply:
Yes, thanks. I think these semantics questions should be answered before we talk about the parallel model.

Reply:
By reading the last code that you wrote:
```
1  vertical_region(k_start, k_end)
2    one = zero
3    two = one[k + 1]
```
I would expect the same thing that I expect when reading horizontal stencil code, that is isolation between statements: line 3 should get executed after line 2 has fully executed, resulting in `two = one = zero`. As comparison, look at the horizontal version of it:
```
1  vertical_region(k_start, k_end)
2    one = zero
3    two = one[i + 1]
```
... since we expect statement isolation we stage-split between 2 and 3.

I don't see the benefit of complicating the intended meaning of `vertical_region`, as you can always obtain the result `one = zero` and `two = one(prev)` by simply writing:
```
1  vertical_region(k_start, k_end)
2    two = one
3    one = zero
```
We gain in simplicity of the analysis and readability of the code if we keep `vertical_region` as just the specifier of the vertical interval in which its statements execute.
What I say we should do to fix the problem of the current implementation, which appears in this case
```
1  vertical_region(k_start, k_end)
2    one = one[k+1]
3    one = one * 2.0
```
(2 and 3 are erroneously put in the same do-method), is to disallow fusion (or introduce a split) when detecting that a statement (line 3) writes on a field accessed vertically offset by a statement that precedes it (line 2). Which is, again, analogous to what we currently do in the horizontal.

Reply:
So what you basically say is that in a vertical region, you want to forbid accessing the "old values", because this can always be done by versioning manually, right? That would mean, that a field, which is read and then written in the same vertical region, it is always a solver. SIR should not have a notion of forward/backward then. The implied solver directions would need to match, so you cannot have a forward read-write dependency and a backward read-write dependency in the same vertical region. 
```
vertical_region(k_start, k_end):
    one = 1
    one = one[k - 1] + one
```
This will be a solver.

```
vertical_region(k_start, k_end):
    one = 1
    one = one[k - 1] + one[k + 1]
```
This is invalid.

Reply:
@lukasm91
> So what you basically say is that in a vertical region, you want to forbid accessing the "old values", because this can always be done by versioning manually, right? 

Yes, because statement-level atomicity would imply that once line 2 (of my last example) has been fully executed, the old values of `one` are lost and line 3 will operate on the new values. But why would you need the old values in any case? Could you provide an example?

> That would mean, that a field, which is read and then written in the same vertical region, it is always a solver. SIR should not have a notion of forward/backward then. The implied solver directions would need to match, so you cannot have a forward read-write dependency and a backward read-write dependency in the same vertical region.

To me having a vertical offset access on a statement implies that the statement is (part of) a vertical solver. Whether we want to switch the order of `k_start` and `k_end` or not in the vertical region declaration, I don't care, because the direction is always clear from the accesses.
The only ambiguous case is when you don't have neither vertical nor horizontal offset accesses:
```
one = one + 1
```
which could be interpreted as either part of a solver or an horizontal stencil. Operationally it shouldn't matter. But we could we could improve DSL code readability by putting vertical solvers inside a `vertical_solver` block.

> 
> ```
> vertical_region(k_start, k_end):
>     one = 1
>     one = one[k - 1] + one
> ```
> 
> This will be a solver.

Yes, and I would code generate it to:
```
for (k=k_start; k<k_end; k++) {
   one(k) = 1;
}
for (k=k_start; k<k_end; k++) {
   one(k) = one(k-1) + one(k);
}
```

> ```
> vertical_region(k_start, k_end):
>     one = 1
>     one = one[k - 1] + one[k + 1]
> ```
> 
> This is invalid.

This would still be valid and I would code generate it to:

```
for (k=k_start; k<k_end; k++) {
   one(k) = 1;
}
for (k=k_start; k<k_end; k++) {
   one(k) = one(k-1) + one(k+1);
}
```

> 
> ```
> vertical_region(k_start, k_end):
>     one = 1
> vertical_region(k_start, k_end):
>     one = one[k - 1] + one[k + 1]
> ```
> 
> This are two 3D stencils (not a solver)

This, to me, it's the exact same as before (same generated code).

Reply:
Vertical offsets are not the same as solvers though. You can also just have normal stencils in the vertical

Reply:
> Vertical offsets are not the same as solvers though. You can also just have normal stencils in the vertical

@twicki right. Then I think it is important to fix the syntax both on the frontend and on the backend as soon as possible.
From what I understand, writing `one = one[k - 1]` in the context of vertical solvers means accumulating, but in the context of stencils means a stencil access (accessing the "old values" of `one` in the rhs, like you would access an array). I can see two alternative solutions:

- Always having vertical solvers in a `vertical_solver` block, where the meaning of `one[k-1]` is that of an array access. Outside a vertical solver, of course, we would mean a stencil access.
- Use a different notation: `one{k-1}` for an array-like access and `one[k-1]` for a stencil access.

Reply:
Let me summarize and see if understood the discussion here. There original problem was with this code:
```
vertical_region(k_start, k_end)
     one = zero
vertical_region(k_start, k_end)
      two = one[k + 1]
```
In the current semantics (I am going to name it **Semantic 0**), any access that is counter loop order, means a stencil (i.e. access old data and not solver).
That is `two = one[k+1]` should be a stencil. The problem is that currently dawn is merging the two vertical regions into 
```
vertical_region(k_start, k_end)
     one = zero
      two = one[k + 1]
```
which clearly shouldnt do since the semantics of the stencil access is to access values written in previous vertical_regions to `one`. But if we fix the merging pass, adding the rule that a solver stmt can not be moved into a vertical_region where we write into the same field, we would fix the problem, and the behaviour would be semantically correct, isnt? 

Then I understood Lukas/Giacomo were proposing to simplify the semantics (**Semantic 1**)by not using the loop order to identify whether accesses are solvers or stencils. Instead we assume that any read access to a field that has been written into in the same vertical_region is a solver. Then the loop order will be defined by the read access of the solver. 
In `out = out[k-1]` should be forward, and the opposite for `out = out[k+1]`. 
Using this semantic, my interpretation diverges though from the examples of Lukas and Giacomo, so here it comes mine:
```
vertical_region(k_start, k_end):
    one = 1
    one = one[k - 1] + one
```
I agree with both here, this is a solver. Though is a strange example since stmt `one=1` is not used in most of the region. I think the following is more readable for me (though it does not change the result)
```
vertical_region(k_start, k_start)::
    one = 1
vertical_region(k_start+1, k_end)::
    one = one[k - 1] + one
```
The example:
```
vertical_region(k_start, k_end):
    one = 1
    one = one[k - 1] + one[k + 1]
```
I agree with @lukasm91, this should be invalid. In principle is a solver but we can not define the loop order of the solver since there are two contradictory accesses. I guess @Stagno was making valid here assuming **Semantic 0** ?

For the example: 
```
vertical_region(k_start, k_end):
    one = 1
vertical_region(k_start, k_end):
    one = one[k - 1] + one[k + 1]
```
I disagree with both, to me this is the way I would actually write a solver, because of the stmt
```    one = one[k - 1] + one[k + 1] ```. Notice that the pre-condition `one = 1` can always go into a separate `vertical_region(k_start,k_start)` as a defined in the example above and is not what defines the solver but the self -update.

Then to conclude we have the possibility of **Semantic 2**, using a `vertical_solver`. I would like to add the caveat that this probably makes difficult to combine solvers accesses (for example to different fields) even in the same statement, which could happen, isnt?
Finally we have the **Semantic 3** of using different notation `{}` an `[]` (i.e. different field access nodes).

Did I get it right? Is there now about choosing or do we still lack some understanding or non covered cases?

Reply:
@cosunae 
In my first comment (https://github.com/MeteoSwiss-APN/dawn/pull/551/files#r366916356) I erroneously assumed that all vertical offset accesses are to be intended as solver accesses. This could be still a viable option because if the user wants a stencil-like access he/she can always manually field-version:

> ```
> vertical_region(k_start, k_end):
>    one = 1;
>    one_out = one[k-1] + one[k+1];
>    one = one_out
> ```

but it's definitely not the one I'd prefer.

In the second comment (https://github.com/MeteoSwiss-APN/dawn/pull/551/files#r366935001), as you said we have Semantic 2 and Semantic 3.
Answering your question on semantic 2: it is important that we thoroughly check if that's a possible use case that we need to support (horizontal stencil + solver access in the same statement). If it is, we can't have semantic 2.

Reply:
> Then I understood Lukas/Giacomo were proposing to simplify the semantics (Semantic 1)by not using the loop order to identify whether accesses are solvers or stencils. Instead we assume that any read access to a field that has been written into in the same vertical_region then is a solver. 
No, I don't think we should do that (my statements were just about trying to understand proposed solutions)! The user always have a direction in mind when he is writing a solver. So we should accept that, and I think there is no reason to reverse engineer what the user might have thought when writing the code.